File: Lecture 3/Python/newton_raphson.py
from create_Yb import create_Yb

import numpy as np
import pickle


def newton_raphson(system):
    """
    Perform power flow analysis using Newton-Raphson's power flow method.
    """

    # Extract the system data
    number_of_buses = system['buses'].shape[0]
    bus_type = system['buses'][:, 1].astype(int)
    pq_buses = np.where(bus_type == 3)[0]
    v_specified = system['buses'][:, 2]
    theta = system['buses'][:, 3]
    p_load = system['buses'][:, 4]
    q_load = system['buses'][:, 5]
    p_gen = system['buses'][:, 6]
    q_gen = system['buses'][:, 7]

    # Calculate the specified active and reactive power injections
    p_specified = p_gen - p_load
    q_specified = q_gen - q_load

    # Create the bus-admittance matrix and evaluate its real and imaginary parts
    Y_b = create_Yb(system)
    G = np.real(Y_b)
    B = np.imag(Y_b)

    # Initialize the calculation variables
    V = np.copy(v_specified)
    V_old = np.zeros_like(V)
    iteration = 0
    tolerance = 1e-6

    # Main loop
    while np.any(np.abs(V * np.exp(1j * theta) - V_old) > tolerance):
        # Update the iteration variables
        V_old = V * np.exp(1j * theta)
        iteration += 1

        # Determine active and reactive power injections
        P = np.zeros(number_of_buses)
        Q = np.zeros(number_of_buses)
        for i in range(number_of_buses):
            for j in range(number_of_buses):
                P[i] += V[i] * V[j] * (G[i, j] * np.cos(theta[i] - theta[j]) + B[i, j] * np.sin(theta[i] - theta[j]))
                Q[i] += V[i] * V[j] * (G[i, j] * np.sin(theta[i] - theta[j]) - B[i, j] * np.cos(theta[i] - theta[j]))

        # Determine the active and reactive power injection deviations
        dP = p_specified - P
        dP = dP[1:]
        dQ = q_specified - Q
        dQ = dQ[pq_buses]

        # Create J1 = dP/dtheta
        J1 = np.zeros((number_of_buses, number_of_buses))
        for i in range(number_of_buses):
            for j in range(number_of_buses):
                if i == j:
                    for k in range(number_of_buses):
                        J1[i, i] += V[i] * V[k] * (
                                    -G[i, k] * np.sin(theta[i] - theta[k]) + B[i, k] * np.cos(theta[i] - theta[k]))
                    J1[i, i] -= V[i] ** 2 * B[i, i]
                else:
                    J1[i, j] = V[i] * V[j] * (
                                G[i, j] * np.sin(theta[i] - theta[j]) - B[i, j] * np.cos(theta[i] - theta[j]))
        J1 = J1[1:, 1:]

        # Create J2 = dP/dV
        J2 = np.zeros((number_of_buses, number_of_buses))
        for i in range(number_of_buses):
            for j in range(number_of_buses):
                if i == j:
                    for k in range(number_of_buses):
                        J2[i, i] += V[k] * (G[i, k] * np.cos(theta[i] - theta[k]) + B[i, k] * np.sin(theta[i] - theta[k]))
                    J2[i, i] += G[i, i] * V[i]
                else:
                    J2[i, j] = V[i] * (G[i, j] * np.cos(theta[i] - theta[j]) + B[i, j] * np.sin(theta[i] - theta[j]))
        J2 = J2[1:, pq_buses]

        # Create J3 = dQ/dtheta
        J3 = np.zeros((number_of_buses, number_of_buses))
        for i in range(number_of_buses):
            for j in range(number_of_buses):
                if i == j:
                    for k in range(number_of_buses):
                        J3[i, i] += V[i] * V[k] * (G[i, k] * np.cos(theta[i] - theta[k]) + B[i, k] * np.sin(theta[i] - theta[k]))
                    J3[i, i] -= G[i, i] * V[i] ** 2
                else:
                    J3[i, j] = -V[i] * V[j] * (G[i, j] * np.cos(theta[i] - theta[j]) + B[i, j] * np.sin(theta[i] - theta[j]))
        J3 = J3[pq_buses, 1:]

        # Create J4 = dQ/dtheta
        J4 = np.zeros((number_of_buses, number_of_buses))
        for i in range(number_of_buses):
            for j in range(number_of_buses):
                if i == j:
                    for k in range(number_of_buses):
                        J4[i, i] += V[k] * (G[i, k] * np.sin(theta[i] - theta[k]) - B[i, k] * np.cos(theta[i] - theta[k]))
                    J4[i, i] -= B[i, i] * V[i]
                else:
                    J4[i, j] = V[i] * (G[i, j] * np.sin(theta[i] - theta[j]) - B[i, j] * np.cos(theta[i] - theta[j]))
        J4 = J4[pq_buses, :][:, pq_buses]

        # Create the complete Jacobian matrix
        J = np.block([[J1, J2], [J3, J4]])

        # Solve the linear system of equations
        dX = np.linalg.solve(J, np.concatenate([dP, dQ]))

        # Update the voltage magnitudes and voltage phase angles
        dTh = dX[:number_of_buses - 1]
        dV = dX[number_of_buses - 1:]
        theta[1:] += dTh
        V[pq_buses] += dV

    # Convert theta to degrees
    theta = np.rad2deg(theta)

    return V, theta, P, Q, iteration


# Sample usage
if __name__ == '__main__':
    # Load the system data
    with open("9 bus system.pkl", "rb") as file:
        system = pickle.load(file)
    # Perform the power flow
    v, theta, P, Q, iteration = newton_raphson(system)
    print("Voltage magnitudes:", v)
    print("Voltage phase angles:", theta)

File: Lecture 3/MATLAB/newton_raphson.m
% This function performs power flow analysis using Newton-Raphson's method
% and it neglects the generator reactive power limits

function [V,Theta,P,Q,Iteration] = newton_raphson(System)

% Extract the bus data and determine the IDs of PV and PQ nodes

Number_of_Buses = size(System.Buses, 1) ;
Bus_Type = System.Buses.Bus_Type ;
PQ_Buses = find(Bus_Type==3) ;

% Extract the specified voltage data

V_specified = System.Buses.V ;
Theta = System.Buses.Theta ;

% Extract the load and generation data

P_load = System.Buses.P_load ;
Q_load = System.Buses.Q_load ;
P_gen = System.Buses.P_gen ;
Q_gen = System.Buses.Q_gen ;

% Calculate the specified active and reactive power injections

P_specified = P_gen - P_load ;
Q_specified = Q_gen - Q_load ;

% Create the bus-admittance matrix and evaluate its magnitude and phase
% angle

Y_b = create_Yb(System) ;
G = real(Y_b) ; 
B = imag(Y_b) ; 

% Initialize the calculation variables

V = V_specified ;
V_old = zeros(size(V)) ;
Iteration = 0 ;
Tolerance = 1e-6 ;

% Main loop

while any(abs(V.*exp(sqrt(-1)*Theta)-V_old)>Tolerance)

    % Update the iteration variables

    V_old = V.*exp(sqrt(-1)*Theta) ;
    Iteration = Iteration + 1 ;

    % Determine active and reactive power injections

    P = zeros(Number_of_Buses,1) ;
    Q = zeros(Number_of_Buses,1) ;

    for i = 1 : Number_of_Buses

        for j = 1 : Number_of_Buses

            P(i) = P(i) + V(i)*V(j)*(G(i,j)*(cos(Theta(i)-Theta(j)))+B(i,j)*sin(Theta(i)-Theta(j))) ;
            Q(i) = Q(i) + V(i)*V(j)*(G(i,j)*(sin(Theta(i)-Theta(j)))-B(i,j)*cos(Theta(i)-Theta(j))) ;

        end

    end

    % Determine the active and reactive power injection deviations

    dP = P_specified - P ;
    dP = dP(2:Number_of_Buses) ;
    dQ = Q_specified - Q ;
    dQ = dQ(PQ_Buses) ;

    % Create the Jacobian submatrix J1 = dP/dTheta

    J1 = zeros(Number_of_Buses) ;

    for i = 1 : Number_of_Buses

        for j = 1 : Number_of_Buses

            if i == j

                for k = 1 : Number_of_Buses

                    J1(i,i) = J1(i,i) + V(i)*V(k)*(-G(i,k)*sin(Theta(i)-Theta(k))+B(i,k)*cos(Theta(i)-Theta(k))) ; 
         
                end

                J1(i,i) = J1(i,i) - V(i)^2*B(i,i) ;

            else

                J1(i,j) = V(i)*V(j)*(G(i,j)*sin(Theta(i)-Theta(j))-B(i,j)*cos(Theta(i)-Theta(j))) ; 

            end

        end

    end

    J1 = J1(2:end,2:end) ;

    % Create the Jacobian submatrix J2 = dP/dV

    J2 = zeros(Number_of_Buses) ;

    for i = 1 : Number_of_Buses

        for j = 1 : Number_of_Buses

            if i == j

                for k = 1 : Number_of_Buses

                    J2(i,i) = J2(i,i) + V(k)*(G(i,k)*cos(Theta(i)-Theta(k))+B(i,k)*sin(Theta(i)-Theta(k))) ;

                end

                J2(i,i) = J2(i,i) + G(i,i)*V(i) ; 

            else

                J2(i,j) = V(i)*(G(i,j)*cos(Theta(i)-Theta(j))+B(i,j)*sin(Theta(i)-Theta(j))) ;

            end

        end

    end

    J2 = J2(2:end,PQ_Buses) ;

    % Create the Jacobian submatrix J3 = dQ/dTheta

    J3 = zeros(Number_of_Buses) ;

    for i = 1 : Number_of_Buses

        for j = 1 : Number_of_Buses

            if i == j

                for k = 1 : Number_of_Buses

                    J3(i,i) = J3(i,i) + V(i)*V(k)*(G(i,k)*cos(Theta(i)-Theta(k))+B(i,k)*sin(Theta(i)-Theta(k))) ;

                end

                J3(i,i) = J3(i,i) - G(i,i)*V(i)^2 ; 

            else

                J3(i,j) = -V(i)*V(j)*(G(i,j)*cos(Theta(i)-Theta(j))+B(i,j)*sin(Theta(i)-Theta(j))) ;

            end

        end

    end

    J3 = J3(PQ_Buses,2:end) ;

    % Create the Jacobian submatrix J4 = dQ/dTheta

    J4 = zeros(Number_of_Buses) ;

    for i = 1 : Number_of_Buses

        for j = 1 : Number_of_Buses

            if i == j

                for k = 1 : Number_of_Buses

                    J4(i,i) = J4(i,i) + V(k)*(G(i,k)*sin(Theta(i)-Theta(k))-B(i,k)*cos(Theta(i)-Theta(k))) ;

                end

                J4(i,i) = J4(i,i) - B(i,i)*V(i) ; 

            else

                J4(i,j) = V(i)*(G(i,j)*sin(Theta(i)-Theta(j))-B(i,j)*cos(Theta(i)-Theta(j))) ;

            end

        end

    end

    J4 = J4(PQ_Buses,PQ_Buses) ;

    % Create the complete Jacobian matrix

    J = [J1 J2 ; J3 J4] ;

    % Solve the linear system of equations

    dX = J\[dP ; dQ] ;

    % Update the voltage magnitudes and voltage phase angles

    dTh = dX(1:Number_of_Buses-1) ;
    dV = dX(Number_of_Buses:end) ;
    Theta(2:Number_of_Buses) = Theta(2:Number_of_Buses) + dTh ;
    V(PQ_Buses) = V(PQ_Buses) + dV ;

end

Theta = Theta*180/pi ;

end